cat/spirv-check.cat

```diff
@@ -0,0 +1,220 @@
+SPIRV
+(* Vulkan + SPIR-V Memory Model *)
+(* This file is not intended for verification purposes, but rather to perform several sanity checks *)
+
+// Base relations:
+// vloc: events accessing the same virtual location, aka "same reference (sref)" in the Vulkan spec
+// syncbar: synchronize by barriers with same logical barrier resource (barID)
+// ssg: same subgroup
+// swg: same workgroup
+// sqf: same queuefamily
+// ssw: system-synchronizes-with
+
+// Tags:
+// SC0, SC1: storage class 0, 1
+// SEMSC0, SEMSC1: storage class semantics 0. 1
+// SG, WG, QF, DV: subgroup, workgroup, queuefamily, device domains
+// AV, VIS: availability, visibility
+// AVDEVICE, VISDEVICE: availability, visibility op to the device domain
+
+(*******************)
+(*  Auxiliaries *)
+(*******************)
+
+(* Tags *)
+let PRIV = (R | W) \ NONPRIV // PRIV is the complement of NONPRIV
+
+let SEMSC01 = SEMSC0 & SEMSC1
+
+let AVSHADER = (AV | SEMAV) & DV
+let VISSHADER = (VIS | SEMVIS) & DV
+
+let AVQF = (AV | SEMAV) & (DV | QF)
+let VISQF = (VIS | SEMVIS) & (DV | QF)
+
+let AVWG = (AV | SEMAV) & (DV | QF | WG)
+let VISWG = (VIS | SEMVIS) & (DV | QF | WG)
+
+let AVSG = AV | SEMAV
+let VISSG = VIS | SEMVIS
+
+(* Relations *)
+let inscope = (DV * DV) |
+              (sqf & ((DV | QF) * (DV | QF))) |
+              (swg & ((DV | QF | WG) * (DV | QF | WG))) |
+              (ssg & ((DV | QF | WG | SG) * (DV | QF | WG | SG)))
+
+let mutordatom = (loc & vloc & inscope & (ATOM * ATOM)) \ id // mutually ordered atomics
+
+let avvisSem = ((SC0 | SC1) * (AVDEVICE | VISDEVICE))
+                | (SC0 * ((SEMAV | SEMVIS) & SEMSC0))
+                | (SC1 * ((SEMAV | SEMVIS) & SEMSC1))
+let avvisSet = [AV | VIS]; (vloc & loc)
+let avvisinc = ((avvisSem | avvisSem^-1) \ id) |
+               (avvisSet | avvisSet^-1)
+
+let posctosem = ((SC0 * SEMSC0) | (SC1 * SEMSC1)) & po
+let posemtosc = ((SEMSC0 * SC0) | (SEMSC1 * SC1)) & po
+
+let asmo = co & (ATOM * ATOM)
+let rs = [REL & ATOM]; ((asmo \ (asmo; asmo+)); [RMW])*
+let hypors = [W & ATOM]; ((asmo \ (asmo; asmo+)); [RMW])*
```

Comment:
Is it true that `((asmo \ (asmo; asmo+)); [RMW])*` is equivalent to `(asmo;[RMW])?`.
Btw. it seems the alloy model assumes rmw's to be single events like RISCV's AMOs. This will cause the read part of the rmw to also be related by `asmo`, `rs`, and `hypors`, which is not the case in the cat-model. I don't know what consequence this has.

Reply:
The first relation guarantees that before making the transitive closure, you only consider immediate `asmo` edges FINISHING in a `RMW`. Thus, after the transitive closure, the whole chain should be over `RMW` events. AFAIS, the relation you propose only guarantees that the final one is `RMW` while all intermediate ones could be `ATOM`.

It is true the alloy model assumes single RMW events. One of the consequences of this is that we explicitly needed to add the atomicity axiom (which is not part of the alloy model).

Strictly speaking, in their model there is no "read part of a RMW" ... but it is true that there will be events tagged by `R` which can start `asmo`, `rs`, `hypors`. However, the same is true for `rf` (they don't really have `co` since `asmo` plays a similar role): you can start by events tagged with `R` and finish in events tagged with `W`. I think this difference happens in most of the alloy-based models which normally use single events for this. E.g., I could not find anything in memalloy [preventing to start a `rf` from an event tagged with `W`](https://github.com/johnwickerson/memalloy/blob/master/archs/exec.als).

However, since all usages of the 3 relations you mentioned are use to guarantee synchronization with some form of `rf` or `co`, I have the feeling that there should not be any difference in practice (but I could be easily wrong about this).

Reply:
You are right about the `amso;[RMW]` thing. I guess my brain wasn't working properly.

I don't know if the two-event rmw modelling has any consequences. Maybe it is effectively equivalent to the original formulation. I just wanted to be sure that you are aware of this difference and that you thought about its correctness.